File: fastapi-backend/parsers/weather.py
"""Погода рядом с СУНЦ НГУ (Академгородок, Новосибирск).

Координаты школы: 54.842° с.ш., 83.098° в.д. (см. анализ, раздел 3.6).

* Основной источник — **Open-Meteo** (бесплатный, без API-ключа): текущая
  погода + прогноз на 3 дня, таймзона Asia/Novosibirsk.
* Резервный — **wttr.in** (``/Akademgorodok?format=j1``): используется при
  сбое или лимитировании Open-Meteo.

Коды погоды WMO (Open-Meteo) и WWO (wttr.in) переводятся в русские описания
и иконки-эмодзи. Направление ветра переводится в румбы (С, СВ, В, …).
"""

from __future__ import annotations

import datetime as dt
from typing import Any

from . import SourceError, http_get_json

#: Координаты СУНЦ НГУ
LATITUDE = 54.842
LONGITUDE = 83.098

#: Основной источник: текущая погода + прогноз на 3 дня (спецификация из анализа, 3.6)
OPEN_METEO_URL = (
    "https://api.open-meteo.com/v1/forecast"
    "?latitude=54.842&longitude=83.098"
    "&current=temperature_2m,relative_humidity_2m,apparent_temperature,"
    "weather_code,wind_speed_10m,wind_direction_10m"
    "&daily=weather_code,temperature_2m_max,temperature_2m_min,"
    "sunrise,sunset,precipitation_probability_max"
    "&timezone=Asia%2FNovosibirsk&forecast_days=3"
)

#: Лёгкий запрос для проверки доступности (эндпоинт /api/health)
HEALTH_URL = (
    "https://api.open-meteo.com/v1/forecast"
    "?latitude=54.842&longitude=83.098&current=temperature_2m&timezone=Asia%2FNovosibirsk"
)

#: Резервный источник
WTTR_URL = "https://wttr.in/Akademgorodok?format=j1"

LOCATION = "Академгородок, Новосибирск (СУНЦ НГУ)"

#: WMO weather code → (русское описание, иконка)
WMO_CODES: dict[int, tuple[str, str]] = {
    0: ("Ясно", "☀️"),
    1: ("Преимущественно ясно", "🌤️"),
    2: ("Малооблачно", "⛅"),
    3: ("Пасмурно", "☁️"),
    45: ("Туман", "🌫️"),
    48: ("Изморозь", "🌫️"),
    51: ("Слабая морось", "🌦️"),
    53: ("Морось", "🌦️"),
    55: ("Сильная морось", "🌧️"),
    56: ("Ледяная морось", "🌧️"),
    57: ("Ледяная морось", "🌧️"),
    61: ("Слабый дождь", "🌦️"),
    63: ("Дождь", "🌧️"),
    65: ("Сильный дождь", "🌧️"),
    66: ("Ледяной дождь", "🌧️"),
    67: ("Ледяной дождь", "🌧️"),
    71: ("Слабый снег", "🌨️"),
    73: ("Снег", "❄️"),
    75: ("Сильный снег", "❄️"),
    77: ("Снежные зёрна", "🌨️"),
    80: ("Слабый ливень", "🌦️"),
    81: ("Ливень", "🌧️"),
    82: ("Сильный ливень", "⛈️"),
    85: ("Снегопад", "🌨️"),
    86: ("Сильный снегопад", "❄️"),
    95: ("Гроза", "⛈️"),
    96: ("Гроза с градом", "⛈️"),
    99: ("Сильная гроза с градом", "⛈️"),
}

#: Коды WWO (wttr.in) → (русское описание, иконка)
WWO_CODES: dict[int, tuple[str, str]] = {
    113: ("Ясно", "☀️"),
    116: ("Малооблачно", "⛅"),
    119: ("Облачно", "☁️"),
    122: ("Пасмурно", "☁️"),
    143: ("Туман", "🌫️"),
    248: ("Туман", "🌫️"),
    260: ("Туман", "🌫️"),
    176: ("Слабый дождь", "🌦️"),
    263: ("Слабый дождь", "🌦️"),
    266: ("Морось", "🌦️"),
    293: ("Слабый дождь", "🌦️"),
    298: ("Дождь", "🌧️"),
    302: ("Дождь", "🌧️"),
    305: ("Сильный дождь", "🌧️"),
    308: ("Ливень", "🌧️"),
    179: ("Слабый снег", "🌨️"),
    323: ("Слабый снег", "🌨️"),
    328: ("Снег", "❄️"),
    332: ("Снег", "❄️"),
    338: ("Сильный снег", "❄️"),
    200: ("Гроза", "⛈️"),
}

#: Румбы розы ветров
WIND_DIRS: tuple[str, ...] = ("С", "СВ", "В", "ЮВ", "Ю", "ЮЗ", "З", "СЗ")

WEEKDAYS_RU: tuple[str, ...] = ("Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс")


def describe_wmo(code: int | None) -> tuple[str, str]:
    """Код WMO → (русское описание, эмодзи)."""
    return WMO_CODES.get(int(code or 0), ("Переменная облачность", "🌡️"))


def describe_wwo(code: int | None) -> tuple[str, str]:
    """Код WWO → (русское описание, эмодзи)."""
    return WWO_CODES.get(int(code or 0), ("Переменная облачность", "🌡️"))


def wind_direction_text(degrees: Any) -> str | None:
    """Градусы направления ветра → румб («СЗ», «ЮВ», …)."""
    if degrees is None:
        return None
    try:
        index = round(float(degrees) % 360 / 45) % 8
    except (TypeError, ValueError):
        return None
    return WIND_DIRS[index]


def _num(value: Any) -> int | float | None:
    """Число из ответа API (int для целых, float для дробных, None)."""
    if value is None:
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return int(number) if number.is_integer() else number


def _time_of_day(iso_value: Any) -> str | None:
    """«2026-08-30T05:56» → «05:56»."""
    if not iso_value:
        return None
    text = str(iso_value)
    return text[11:16] if len(text) >= 16 else text


def _am_pm_to_24h(value: Any) -> str | None:
    """«5:56 AM» (wttr.in) → «05:56»."""
    if not value:
        return None
    try:
        parsed = dt.datetime.strptime(str(value).strip(), "%I:%M %p")
        return parsed.strftime("%H:%M")
    except ValueError:
        return str(value)


def _weekday_of(date_iso: str | None) -> str | None:
    """«2026-08-30» → «Вс»."""
    if not date_iso:
        return None
    try:
        return WEEKDAYS_RU[dt.date.fromisoformat(str(date_iso)[:10]).weekday()]
    except ValueError:
        return None


def _seq(daily: dict[str, Any], key: str, index: int) -> Any:
    """Безопасно взять элемент ``index`` из массива ``daily[key]``."""
    values = daily.get(key)
    if isinstance(values, list) and index < len(values):
        return values[index]
    return None


# ---------------------------------------------------------------------------
# Open-Meteo (основной источник)
# ---------------------------------------------------------------------------

async def _open_meteo(client: Any) -> dict[str, Any]:
    """Текущая погода и прогноз на 3 дня от Open-Meteo."""
    data = await http_get_json(client, OPEN_METEO_URL, timeout=15.0)
    current = data.get("current") or {}
    daily = data.get("daily") or {}

    code = _num(current.get("weather_code"))
    description, emoji = describe_wmo(int(code) if code is not None else 0)
    current_payload = {
        "temperature": _num(current.get("temperature_2m")),
        "apparent": _num(current.get("apparent_temperature")),
        "humidity": _num(current.get("relative_humidity_2m")),
        "wind_speed": _num(current.get("wind_speed_10m")),
        "wind_direction": _num(current.get("wind_direction_10m")),
        "wind_direction_text": wind_direction_text(current.get("wind_direction_10m")),
        "code": int(code) if code is not None else None,
        "description": description,
        "emoji": emoji,
    }

    forecast: list[dict[str, Any]] = []
    dates = daily.get("time") or []
    for index, date_iso in enumerate(dates[:3]):
        day_code = _num(_seq(daily, "weather_code", index))
        day_description, day_emoji = describe_wmo(int(day_code) if day_code is not None else 0)
        forecast.append(
            {
                "date": str(date_iso),
                "weekday": _weekday_of(str(date_iso)),
                "description": day_description,
                "emoji": day_emoji,
                "temp_max": _num(_seq(daily, "temperature_2m_max", index)),
                "temp_min": _num(_seq(daily, "temperature_2m_min", index)),
                "precipitation_probability": _num(_seq(daily, "precipitation_probability_max", index)),
                "sunrise": _time_of_day(_seq(daily, "sunrise", index)),
                "sunset": _time_of_day(_seq(daily, "sunset", index)),
            }
        )

    return {
        "source": "open-meteo",
        "location": LOCATION,
        "current": current_payload,
        "forecast": forecast,
    }


# ---------------------------------------------------------------------------
# wttr.in (резервный источник)
# ---------------------------------------------------------------------------

async def _wttr(client: Any) -> dict[str, Any]:
    """Текущая погода и прогноз от wttr.in (формат j1)."""
    data = await http_get_json(client, WTTR_URL, timeout=20.0)
    condition = (data.get("current_condition") or [{}])[0]

    code = _num(condition.get("weatherCode"))
    description, emoji = describe_wwo(int(code) if code is not None else 0)
    if description == "Переменная облачность":
        english = ""
        descriptions = condition.get("weatherDesc") or []
        if descriptions and isinstance(descriptions[0], dict):
            english = str(descriptions[0].get("value") or "")
        if english:
            description = english

    current_payload = {
        "temperature": _num(condition.get("temp_C")),
        "apparent": _num(condition.get("FeelsLikeC")),
        "humidity": _num(condition.get("humidity")),
        "wind_speed": _num(condition.get("windspeedKmph")),
        "wind_direction": _num(condition.get("winddirDegree")),
        "wind_direction_text": wind_direction_text(condition.get("winddirDegree")),
        "code": int(code) if code is not None else None,
        "description": description,
        "emoji": emoji,
    }

    forecast: list[dict[str, Any]] = []
    for day in (data.get("weather") or [])[:3]:
        hourly = day.get("hourly") or []
        midday = hourly[len(hourly) // 2] if hourly else {}
        day_code = _num(midday.get("weatherCode"))
        day_description, day_emoji = describe_wwo(int(day_code) if day_code is not None else 0)
        astronomy = (day.get("astronomy") or [{}])[0]
        precip_prob = None
        if hourly:
            try:
                precip_prob = max(int(h.get("chanceofrain", 0)) for h in hourly)
            except (TypeError, ValueError):
                precip_prob = None
        date_iso = str(day.get("date") or "")
        forecast.append(
            {
                "date": date_iso,
                "weekday": _weekday_of(date_iso),
                "description": day_description,
                "emoji": day_emoji,
                "temp_max": _num(day.get("maxtempC")),
                "temp_min": _num(day.get("mintempC")),
                "precipitation_probability": precip_prob,
                "sunrise": _am_pm_to_24h(astronomy.get("sunrise")),
                "sunset": _am_pm_to_24h(astronomy.get("sunset")),
            }
        )

    return {
        "source": "wttr.in",
        "location": LOCATION,
        "current": current_payload,
        "forecast": forecast,
    }


# ---------------------------------------------------------------------------
# Публичная функция
# ---------------------------------------------------------------------------

async def get_weather(client: Any) -> dict[str, Any]:
    """Погода: Open-Meteo, при сбое — wttr.in.

    :return: словарь-поле ответа ``GET /api/weather`` (без ``stale``);
    :raises SourceError: если недоступны оба источника.
    """
    open_meteo_error: Exception | None = None
    try:
        return await _open_meteo(client)
    except Exception as exc:  # noqa: BLE001 — переходим к резервному источнику
        open_meteo_error = exc
    try:
        return await _wttr(client)
    except Exception as exc:  # noqa: BLE001
        raise SourceError(f"Open-Meteo: {open_meteo_error}; wttr.in: {exc}") from exc

File: src/lib/server/weather.ts
/**
 * Погода в Академгородке (раздел 3.6 документа).
 * Основной источник — Open-Meteo (без ключа), резервный — wttr.in.
 */

import { SESC_COORDS, fetchWithTimeout } from "./sources";

export interface WeatherCurrent {
  temperature: number | null;
  apparent: number | null;
  humidity: number | null;
  windSpeed: number | null;
  windDirection: string | null;
  description: string;
  icon: string;
  sunrise: string | null;
  sunset: string | null;
}

export interface WeatherForecastDay {
  day: string;
  date: string;
  icon: string;
  description: string;
  tempMax: number | null;
  tempMin: number | null;
  precipitationProbability: number | null;
}

export interface WeatherData {
  current: WeatherCurrent;
  forecast: WeatherForecastDay[];
  source: string;
}

/** WMO weather code → [описание, эмодзи] */
export function describeWmo(code: number | null): [string, string] {
  if (code === null) return ["—", "🌡"];
  const map: Record<number, [string, string]> = {
    0: ["Ясно", "☀️"],
    1: ["Преимущественно ясно", "🌤"],
    2: ["Малооблачно", "⛅"],
    3: ["Пасмурно", "☁️"],
    45: ["Туман", "🌫"],
    48: ["Изморозь", "🌫"],
    51: ["Слабая морось", "🌦"],
    53: ["Морось", "🌦"],
    55: ["Сильная морось", "🌧"],
    56: ["Ледяная морось", "🌧"],
    57: ["Сильная ледяная морось", "🌧"],
    61: ["Небольшой дождь", "🌦"],
    63: ["Дождь", "🌧"],
    65: ["Сильный дождь", "🌧"],
    66: ["Ледяной дождь", "🌧"],
    67: ["Сильный ледяной дождь", "🌧"],
    71: ["Небольшой снег", "🌨"],
    73: ["Снег", "❄️"],
    75: ["Сильный снег", "❄️"],
    77: ["Снежные зёрна", "🌨"],
    80: ["Ливни", "🌦"],
    81: ["Ливни", "🌧"],
    82: ["Сильные ливни", "⛈"],
    85: ["Снежные ливни", "🌨"],
    86: ["Сильные снежные ливни", "❄️"],
    95: ["Гроза", "⛈"],
    96: ["Гроза с градом", "⛈"],
    99: ["Сильная гроза с градом", "⛈"],
  };
  return map[code] ?? ["Переменная облачность", "⛅"];
}

/** WWO weather code (wttr.in) → [описание, эмодзи] */
export function describeWwo(code: number | null): [string, string] {
  if (code === null) return ["—", "🌡"];
  const map: Record<number, [string, string]> = {
    113: ["Ясно", "☀️"],
    116: ["Малооблачно", "⛅"],
    119: ["Пасмурно", "☁️"],
    122: ["Пасмурно", "☁️"],
    143: ["Туман", "🌫"],
    176: ["Небольшой дождь", "🌦"],
    200: ["Гроза", "⛈"],
    227: ["Небольшой снег", "🌨"],
    230: ["Сильный снег", "❄️"],
    248: ["Туман", "🌫"],
    260: ["Туман с изморозью", "🌫"],
    263: ["Слабая морось", "🌦"],
    266: ["Морось", "🌦"],
    281: ["Ледяная морось", "🌧"],
    284: ["Сильная ледяная морось", "🌧"],
    293: ["Небольшой дождь", "🌦"],
    296: ["Дождь", "🌧"],
    299: ["Сильный дождь", "🌧"],
    302: ["Дождь", "🌧"],
    305: ["Сильный дождь", "🌧"],
    308: ["Очень сильный дождь", "⛈"],
    311: ["Ледяной дождь", "🌧"],
    314: ["Ледяной дождь", "🌧"],
    317: ["Снег с дождём", "🌨"],
    320: ["Снег", "❄️"],
    323: ["Небольшой снег", "🌨"],
    326: ["Небольшой снег", "🌨"],
    329: ["Снег", "❄️"],
    332: ["Снег", "❄️"],
    335: ["Сильный снег", "❄️"],
    338: ["Сильный снег", "❄️"],
    350: ["Ледяная морось", "🌧"],
    353: ["Ливни", "🌦"],
    356: ["Сильные ливни", "🌧"],
    359: ["Очень сильные ливни", "⛈"],
    362: ["Снежные ливни", "🌨"],
    365: ["Сильные снежные ливни", "❄️"],
    368: ["Небольшой снег", "🌨"],
    371: ["Снег", "❄️"],
    374: ["Снежные ливни", "🌨"],
    377: ["Снежные зёрна", "🌨"],
    386: ["Гроза с дождём", "⛈"],
    389: ["Сильная гроза", "⛈"],
    392: ["Гроза со снегом", "⛈"],
    395: ["Сильная гроза со снегом", "⛈"],
  };
  return map[code] ?? ["Переменная облачность", "⛅"];
}

function windDirectionText(degrees: number | null): string | null {
  if (degrees === null) return null;
  const dirs = ["С", "СВ", "В", "ЮВ", "Ю", "ЮЗ", "З", "СЗ"];
  return dirs[Math.round(degrees / 45) % 8];
}

const WEEKDAY_NAMES = ["Вс", "Пн", "Вт", "Ср", "Чт", "Пт", "Сб"];

/** Open-Meteo — основной источник */
async function openMeteo(): Promise<WeatherData> {
  const url =
    `https://api.open-meteo.com/v1/forecast?latitude=${SESC_COORDS.lat}&longitude=${SESC_COORDS.lon}` +
    "&current=temperature_2m,relative_humidity_2m,apparent_temperature,weather_code,wind_speed_10m,wind_direction_10m" +
    "&daily=weather_code,temperature_2m_max,temperature_2m_min,sunrise,sunset,precipitation_probability_max" +
    "&timezone=Asia%2FNovosibirsk&forecast_days=3";
  const response = await fetchWithTimeout(url, { timeoutMs: 15000 });
  if (!response.ok) throw new Error(`Open-Meteo вернул HTTP ${response.status}`);
  const data = (await response.json()) as {
    current?: Record<string, number>;
    daily?: Record<string, Array<number | string>>;
  };

  const code = data.current?.weather_code ?? null;
  const [description, icon] = describeWmo(code);
  const current: WeatherCurrent = {
    temperature: data.current?.temperature_2m ?? null,
    apparent: data.current?.apparent_temperature ?? null,
    humidity: data.current?.relative_humidity_2m ?? null,
    windSpeed: data.current?.wind_speed_10m ?? null,
    windDirection: windDirectionText(data.current?.wind_direction_10m ?? null),
    description,
    icon,
    sunrise: data.daily?.sunrise?.[0] ? String(data.daily.sunrise[0]).slice(11, 16) : null,
    sunset: data.daily?.sunset?.[0] ? String(data.daily.sunset[0]).slice(11, 16) : null,
  };

  const forecast: WeatherForecastDay[] = [];
  const daily = data.daily ?? {};
  const dates = (daily.time ?? []) as string[];
  for (let i = 0; i < Math.min(3, dates.length); i++) {
    const dayCode = (daily.weather_code ?? [])[i] as number | undefined;
    const [desc, dayIcon] = describeWmo(dayCode ?? null);
    const dateObj = new Date(dates[i] + "T00:00:00");
    forecast.push({
      day: i === 0 ? "Сегодня" : WEEKDAY_NAMES[dateObj.getDay()] ?? "",
      date: dates[i].split("-").reverse().join("."),
      icon: dayIcon,
      description: desc,
      tempMax: (daily.temperature_2m_max ?? [])[i] as number | undefined ?? null,
      tempMin: (daily.temperature_2m_min ?? [])[i] as number | undefined ?? null,
      precipitationProbability:
        ((daily.precipitation_probability_max ?? [])[i] as number | undefined) ?? null,
    });
  }

  return { current, forecast, source: "Open-Meteo" };
}

/** wttr.in — резервный источник */
async function wttrIn(): Promise<WeatherData> {
  const response = await fetchWithTimeout(
    "https://wttr.in/Akademgorodok?format=j1",
    { timeoutMs: 20000, headers: { Accept: "application/json" } }
  );
  if (!response.ok) throw new Error(`wttr.in вернул HTTP ${response.status}`);
  const data = (await response.json()) as {
    current_condition?: Array<Record<string, string>>;
    weather?: Array<{
      date?: string;
      mintempC?: string;
      maxtempC?: string;
      hourly?: Array<{ chanceofrain?: string; weatherCode?: string }>;
      astronomy?: Array<{ sunrise?: string; sunset?: string }>;
    }>;
  };

  const now = data.current_condition?.[0] ?? {};
  const code = Number(now.weatherCode) || null;
  const [description, icon] = describeWwo(code);

  const to24h = (value: string | undefined): string | null => {
    if (!value) return null;
    const m = value.match(/(\d{1,2}):(\d{2})\s*(AM|PM)/i);
    if (!m) return value;
    let hours = Number(m[1]);
    if (m[3].toUpperCase() === "PM" && hours !== 12) hours += 12;
    if (m[3].toUpperCase() === "AM" && hours === 12) hours = 0;
    return `${String(hours).padStart(2, "0")}:${m[2]}`;
  };

  const current: WeatherCurrent = {
    temperature: Number(now.temp_C) || null,
    apparent: Number(now.FeelsLikeC) || null,
    humidity: Number(now.humidity) || null,
    windSpeed: Number(now.windspeedKmph) || null,
    windDirection: now.winddir16Point ?? null,
    description,
    icon,
    sunrise: to24h(data.weather?.[0]?.astronomy?.[0]?.sunrise),
    sunset: to24h(data.weather?.[0]?.astronomy?.[0]?.sunset),
  };

  const forecast: WeatherForecastDay[] = [];
  for (let i = 0; i < Math.min(3, data.weather?.length ?? 0); i++) {
    const day = data.weather![i];
    const dayCode = Number(day.hourly?.[4]?.weatherCode) || null;
    const [desc, dayIcon] = describeWwo(dayCode);
    const chance = Math.max(
      ...(day.hourly ?? []).map((h) => Number(h.chanceofrain) || 0),
      0
    );
    const dateObj = day.date ? new Date(day.date + "T00:00:00") : new Date();
    forecast.push({
      day: i === 0 ? "Сегодня" : WEEKDAY_NAMES[dateObj.getDay()] ?? "",
      date: (day.date ?? "").split("-").reverse().join("."),
      icon: dayIcon,
      description: desc,
      tempMax: Number(day.maxtempC) || null,
      tempMin: Number(day.mintempC) || null,
      precipitationProbability: chance || null,
    });
  }

  return { current, forecast, source: "wttr.in" };
}

/** Погода: Open-Meteo, при сбое — wttr.in */
export async function getWeather(): Promise<WeatherData> {
  try {
    return await openMeteo();
  } catch (openMeteoError) {
    try {
      return await wttrIn();
    } catch (wttrError) {
      throw new Error(
        `Погодные источники недоступны: Open-Meteo — ${(openMeteoError as Error).message}; wttr.in — ${(wttrError as Error).message}`
      );
    }
  }
}
